### ISSUE TITLE
dlopen failed: cannot locate symbol (java.lang.UnsatisfiedLinkError)

### ISSUE THREAD

**Issue by Gibson85** [android/ndk#1185]
Hi, have problems linking multiple shared libs (with circular dependencies) against my native android app/lib.

My CMake looks like this:
```
set(MyLibA "MyLibA")
add_library(${MyLibA} SHARED IMPORTED)
set_target_properties(${MyLibA} PROPERTIES IMPORTED_LOCATION "Path/To/The/Lib/libMyLibA.so")

...

target_link_libraries(
		MyNativeApp
		PUBLIC
		android
		${log-lib}
        EGL
		GLESv1_CM
		GLESv2
		${MyLibA}
		${MyLibB}
		${MyLibC}
		${MyLibD}
		${MyLibE}
)
```
Build.gradle:
```
android {
    compileSdkVersion 29
    defaultConfig {
        applicationId "com.example.mynativeapp"
        minSdkVersion 18
        targetSdkVersion 29
        versionCode 1
        versionName "1.0"
        externalNativeBuild {
            cmake {
                arguments '-DANDROID_STL=c++_shared', '-DANDROID_TOOLCHAIN=clang', '-DANDROID_ALLOW_UNDEFINED_SYMBOLS=ON'
            }
        }
        ndk {
            abiFilters 'armeabi-v7a'//, 'arm64-v8a', 'x86', 'x86_64'
        }
    }
    buildTypes {
        debug {
            minifyEnabled false
            shrinkResources false
            //proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
        release {
            minifyEnabled false
            shrinkResources false
            //proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
    }
    sourceSets {
        main {
            jniLibs.srcDirs = ['libs', 'src/debug/jniLibs', 'src/main/jniLibs', 'src/release/jniLibs']
        }
    }
```

```
    --------- beginning of crash
2020-02-09 00:17:55.656 8275-8275/com.example.mynativeapp E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.example.mynativeapp, PID: 8275
    java.lang.UnsatisfiedLinkError: Unable to load native library "/data/app/com.example.mynativeapp-2/lib/arm/libMyNativeApp.so": dlopen failed: cannot locate symbol "_ZN6...6...10...ERKNS_6...E" referenced by "/data/app/com.example.mynativeapp-2/lib/arm/libMyLibB.so"...
        at android.app.NativeActivity.onCreate(NativeActivity.java:182)
        at android.app.Activity.performCreate(Activity.java:6684)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1119)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2637)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2751)
        at android.app.ActivityThread.-wrap12(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1496)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6186)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
```
I also checked the shared lib. The "missing" symbol is still existing. If I open the shared lib per dlopen on Linux...

```
void loadLibrary(const char* file)
{
	void* handle = dlopen("/media/user/DRIVE/Path/MyLibB.so", RTLD_NOW);
	if (handle != nullptr)
	{
		fprintf(stderr, "Success: %s\n", dlerror());
	}
	else
	{
		fprintf(stderr, "Error: %s\n", dlerror());
	}
}
```
... I don't get a error message. So why android can't link these libs? Any ideas what I could check to find the reason?

**Comment by alexcohn** [comment]
Try to load the library that you doubt about in Android with `loadLibrary()`, i.e.
````
System.loadLibrary("MyLibB");
````

**Comment by Gibson85** [comment]
Thanks for your quick response. I did so using the "hello-libs" NDK example, but it gives me the same error:

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.example.hellolibs, PID: 12538
    java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "_ZN6...6...10...ERKNS_6...E" referenced by "/data/app/com.example.hellolibs-1/lib/arm/libMyLibB.so"...
        at java.lang.Runtime.loadLibrary0(Runtime.java:989)
        at java.lang.System.loadLibrary(System.java:1562)
        at com.example.hellolibs.MainActivity.<clinit>(MainActivity.java:38)
        at java.lang.Class.newInstance(Native Method)
        at android.app.Instrumentation.newActivity(Instrumentation.java:1079)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2576)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2751)
        at android.app.ActivityThread.-wrap12(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1496)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6186)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
```
Anything else I could try?

**Comment by Gibson85** [comment]
Is there a way I could extract the libs packed into the apk file to see if the symbol was maybe stripped away later? There is an apk inspector but there seems not to be an export option. Or maybe it could be a name mangling issue? The symbol that was not found is a class member function. I checked the existing of the function symbol with (readelf -Ws libMyLibB.so).

**Comment by Gibson85** [comment]
Ah maybe I found the reason.
The libMyLibB.so in the "build/" directory contains the function symbol but it is declared as "NOTYPE" ...
```
    13: 00000000     0 NOTYPE  GLOBAL DEFAULT  UND _ZN6...6...10...ERKNS_6...E
```
...while the function symbol of the original lib looks like this:
```
1020348: 005051cd  2400 FUNC    GLOBAL DEFAULT   13 _ZN6...6...10...ERKNS_6...E
```
Could this be an error caused by the stripping?

**Comment by alexcohn** [comment]
The libraries can be extracted from an APK with simple `unzip`. 

But I am afraid I don't understand what you found about **libMyLibB.so**. Please try to load this library alone, because now your findings may be obscured by the other libraries involved.

And I hope the missing symbol has some meaningful characters where you show dots.

**Comment by Gibson85** [comment]
It's getting crazier and crazier.
Today I compiled everything again and suddently without having changed anything all linking problems are gone. I also unzipped the generated apk file's libs and the missing symbol was existing (now). After that I run my app once, but sadly still the log message in my "ANativeActivity_onCreate" function did not appear in logcat (I am not using android_native_app_glue). Then I did a second run and suddently the message appeared! But after the third and forth run not again.

So my guess was that at the end the code was executed well all the time, but maybe the log message was not logged properly by Android Studio. After having run my app for about 100 times now I can say that the log messages appeared for only about 3 times. But you can imagine that this is completely unacceptable. Even break points are not getting hit then at all. There's something completely strange in that very early program startup phase. All my app is doing for now is printing the following log message:

```
JNIEXPORT void ANativeActivity_onCreate(ANativeActivity* activity, void* savedState, size_t savedStateSize)
{
	LOGV("ANativeActivity_onCreate: %p...\n", reinterpret_cast<void*>(activity));
}
```

One other thing I recognized was that when having copied my latest compiled libs to the android app directory even after a build or "Refresh linked C++ Projects" the apk was not updated at all. So at least now I now I always have to do a complete rebuild...

So what I would like to know:

1.) Is there a way to ensure that the log message gets logged? Is there maybe a flush function for that purpose? I guess "__android_log_print" will buffer the messages on first step.
2.) If I set the "android:hasCode=false" property in my manifest file I see a...

 "W/System: ClassLoader referenced unknown path: "

...message in my logcat with an obvious empty path. What does that mean?

**Comment by alexcohn** [comment]
Early logcat messages may be hidden in the Android Studio window if you
choose "only this app" filter. Try to set your filter by your TAG, or even
better – run `adb logcat` on command line and analyse the results off-line.

>
Regarding breakpoints at this very early stage – they aren't reliable. We
put a loop with sleep that polls some global variable before the first
breakpoint that we set. Then we start the app, suspend it when it finally
gets connected to debugger, change the global variable to `false`, and
resume the execution.

**Comment by Gibson85** [comment]
Here's a full thread dump after startup. Can you see something unusual here?

```
"main@4375" prio=5 tid=0x1 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at android.app.NativeActivity.loadNativeCode(NativeActivity.java:-1)
	  at android.app.NativeActivity.onCreate(NativeActivity.java:175)
	  at android.app.Activity.performCreate(Activity.java:6684)
	  at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1119)
	  at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2637)
	  at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2751)
	  at android.app.ActivityThread.-wrap12(ActivityThread.java:-1)
	  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1496)
	  at android.os.Handler.dispatchMessage(Handler.java:102)
	  at android.os.Looper.loop(Looper.java:154)
	  at android.app.ActivityThread.main(ActivityThread.java:6186)
	  at java.lang.reflect.Method.invoke(Method.java:-1)
	  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
	  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)

"FinalizerDaemon@4385" daemon prio=5 tid=0x1aa0 nid=NA waiting
  java.lang.Thread.State: WAITING
	 blocks FinalizerDaemon@4385
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Object.wait(Object.java:407)
	  at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:188)
	  - locked <0x1125> (a java.lang.Object)
	  at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:209)
	  at java.lang.Daemons$FinalizerDaemon.run(Daemons.java:204)
	  at java.lang.Thread.run(Thread.java:761)

"FinalizerWatchdogDaemon@4383" daemon prio=5 tid=0x1aa1 nid=NA waiting
  java.lang.Thread.State: WAITING
	 blocks FinalizerWatchdogDaemon@4383
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Daemons$FinalizerWatchdogDaemon.sleepUntilNeeded(Daemons.java:269)
	  - locked <0x1124> (a java.lang.Daemons$FinalizerWatchdogDaemon)
	  at java.lang.Daemons$FinalizerWatchdogDaemon.run(Daemons.java:249)
	  at java.lang.Thread.run(Thread.java:761)

"ReferenceQueueDaemon@4382" daemon prio=5 tid=0x1a9f nid=NA waiting
  java.lang.Thread.State: WAITING
	 blocks ReferenceQueueDaemon@4382
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Daemons$ReferenceQueueDaemon.run(Daemons.java:150)
	  - locked <0x386> (a java.lang.Class)
	  at java.lang.Thread.run(Thread.java:761)

"EGL Init@4377" prio=5 tid=0x1aa5 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at android.opengl.EGL14.eglGetDisplay(EGL14.java:-1)
	  at android.os.GraphicsEnvironment.-android_os_GraphicsEnvironment_lambda$1(GraphicsEnvironment.java:47)
	  at android.os.GraphicsEnvironment$-void_setupGraphicsEnvironment_android_content_Context_context_LambdaImpl0.run(GraphicsEnvironment.java:-1)
	  at java.lang.Thread.run(Thread.java:761)

"HeapTaskDaemon@4384" daemon prio=5 tid=0x1aa2 nid=NA waiting for monitor entry
  java.lang.Thread.State: BLOCKED
	  at dalvik.system.VMRuntime.runHeapTasks(VMRuntime.java:-1)
	  at java.lang.Daemons$HeapTaskDaemon.run(Daemons.java:433)
	  at java.lang.Thread.run(Thread.java:761)

"Jit thread pool worker thread 0@4380" daemon prio=5 tid=0x1a9c nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	 Incompatible thread state: thread not suspended

"Signal Catcher@4381" daemon prio=5 tid=0x1a9d nid=NA waiting
  java.lang.Thread.State: WAITING
	 Incompatible thread state: thread not suspended

"Binder:15401_1@4386" prio=5 tid=0x1aa3 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	 Incompatible thread state: thread not suspended

"Binder:15401_2@4387" prio=5 tid=0x1aa4 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	 Incompatible thread state: thread not suspended
```

**Comment by enh-google** [comment]
> Regarding breakpoints at this very early stage – they aren't reliable. We put a loop with sleep that polls some global variable before the first breakpoint that we set. Then we start the app, suspend it when it finally gets connected to debugger, change the global variable to `false`, and resume the execution.

i asked the Studio folks about this, and it's news to them. (they also couldn't reproduce with a trivial example.) if you have repro steps, file a bug with them (https://developer.android.com/studio/report-bugs). if you let me know the bug id i'll make sure the right folks get to see it.

**Comment by alexcohn** [comment]
@enh-google I don't have a working example of unreliable breakpoints at hand, I am no longer working for that company. And quite possibly, this happened only on _some_ devices, and not with the latest versions of AS and SDK. 

My comment was not a form of bug report. My point was that if AS doesn't show something in logcat window or stop at some breakpoint, it does not necessarily mean that this code is not executed.

**Comment by alexcohn** [comment]
> Here's a full thread dump after startup. Can you see something unusual here?

why should there be anything unusual?

**Comment by Gibson85** [comment]
@enh-google
I am fighting with these annoying problems since weeks now.
So do me a favour. Take your android studio folks and kick them into their ass from me.
I can't believe what a saw now.

I thought the debugger would be broken completely, because the lines (e.g. at NativeActivity.java) I had put the breakpoints on, had been greyed out because there would be no executable code. "Luckily" I only found these greyed out lines. Now I did a new try and there are really lines you can put a "working" breakpoint on. By the way there are IDEs that will move the breakpoint to the next line that can be executed. But thats not the problem at all. I placed now three breakpoints that not had been greyed out and started debugging. All these three breakpoints have a checkmark also. But the first breakpoint...

NativeActivity.java line 160:
```
String path = classLoader.findLibrary(libname);
```
...never gets hit at all.
Ok, but there are two other chances left. So lets hope the best.
The second breakpoint at line 167...
```
byte[] nativeSavedState = savedInstanceState != null ...
```
...really gets hit.
Originally I was interested in line 170 and the return value, but as said before the breakpoint was greyed out. So finally the third breakpoint at line 176 was hit also and showed that the "mNativeHandle" was zero. Ok, that was unexpected to me. Then on next line 177 the "UnsatisfiedLinkError" exception was thrown claiming it couldn't find my app's shared lib. Fun part. This exception never was logged in logcat and even crazier although an exception was thrown the next line 180
```
super.onCreate(savedInstanceState);
```
was executed nevertheless.

And that's not all.

I have to control myself very much now.
In this fuc.... logcat window now all these log messages I wanted to see since weeks appeared. Most probably because your shitty Android is completely broken with maybe heavy multithreading or other timing problems. If I step through the code with the debugger all these messages appear now but not by running it!

BUT THATS STILL NOT ALL!

Before I started these debugger sessions I commented out this fuc... "ANativeActivity_onCreate" function and even veryfied by readelf that there is not a single "ANativeActivity_onCreate" symbol at any lib I was linking to. I also verified that I had activated the linker flags that no missing symbols are allowed. In addition I did a project cleanup, refreshed linked C++ piojects and did a complete rebuild not only once I did it twice.

BUT STILL I WAS GETTING THESE LOG MESSAGES IN THE ANDROID STUDIOS LOCAT WINDOW FROM A FUNCTION NOT BEING EXISTEND ANYMORE.

Finally I removed all android build files manually and believe it or not now I got what I would have expected before:
```
2020-02-11 18:40:52.026 22401-22401/com.mtk.spacefighters W/NativeActivity: ANativeActivity_onCreate not found
```
But again. This log messages only appear when stepping through the code with the debugger. If you "Run" the code there is nothing!

THERE IS NOTHING!

I hope this is trivial enough for you and your android folks to find out the bugs.
And I hope this will be fixed soon. And by the way. If that behavior gets proven by you folks,
I think the least you could do to me is paying a monthly salary as compensation :D.

I wish I would not have get confused by the greyed out breakpoints respectively by breakpoint not getting hit at all. That could have saved my a lot of time because I was completely blind on "debugging"...

To be honest. Please fix these annoying problems. Please!

To sum it up again:
 - Android studio eats your log messages not being logged at all.
 - Android studio eats your breakpoints.
 - Android studio links together some non existing code. Do you remember the "InstaRun" desaster. I thought this never woking sh.. was removed?
 - Android studio often doesn't update your apk even if changes had been made that would require a new build.

**Comment by DanAlbert** [comment]
Please file bugs against Android Studio. Like @enh-google said, we can help get them triaged, but we're not the people that work on those products.

Some advice on getting results from your bug reports when you file them: break that up into a separate report for each issue. If one report describes several issues it's difficult to find the individual issues and it's difficult to make sure that each part is properly assigned. Make sure there are instructions for reproducing each issue (a sample project is best). If the problem can't be reproduced by the engineers working on that project there isn't much they can do to fix it.

Paste the links here when you file them and we can make sure they get assigned quickly.

**Comment by Gibson85** [comment]
That sounds reasonable. I will do some further investigations and open the reports at the weekend.

**Comment by DanAlbert** [comment]
There's a lot in here, so perhaps I missed it, but it looks like we weren't able to find a concrete NDK issue here? Is there still investigation to do? Seems the answer is no, so closing. lmk if I missed something and we should reopen.

**Comment by alexcohn** [comment]
> i asked the Studio folks about this, and it's news to them.

@enh-google Actually, it's very strange that this is new for them: see https://developer.android.com/ndk/guides/wrap-script:

> When debugging and profiling apps with native code, it’s often useful to use debugging tools that need to be enabled at process startup.

**Comment by enh-google** [comment]
> > i asked the Studio folks about this, and it's news to them.
> 
> @enh-google Actually, it's very strange that this is new for them: see https://developer.android.com/ndk/guides/wrap-script:
> 
> > When debugging and profiling apps with native code, it’s often useful to use debugging tools that need to be enabled at process startup.

what you're missing there is that _i_ wrote that documentation, and i've never used Studio to debug anything. i'm a command-line dinosaur who honestly uses strace more than he uses a debugger. even when i do use a command-line debugger, it's usually still gdb. hopefully @hhb will help me switch to command-line lldb this year though...

(this is an unfortunate problem both inside and outside our organization --- there's relatively little overlap between command-line and GUI users. neither tends to know much about the other.)

**Comment by alexcohn** [comment]
So the bottom line is, the problem of unreliable debugging and logcat early on app start is not specific to Android Studio, but is inherent in the platform. One thing is sure, it's no way an NDK fault.

**Comment by enh-google** [comment]
> So the bottom line is, the problem of unreliable debugging and logcat early on app start is not specific to Android Studio, but is inherent in the platform. One thing is sure, it's no way an NDK fault.

i think it's more "we are not aware of any currently reproducible issues". but, yeah, even if anyone is they're probably not NDK issues. but we can at least try to help find the right Studio people for Studio issues, and sometimes we're also (or sit next to) the right platform person.

but the comments on the wrap-script page are really just "the command-line is also an option".

**Comment by alexcohn** [comment]
_"the command-line is also an option"_ absolutely, but more than that: issues with app load and start may **require** command-line debugging or something equivalent (I sometimes put an artificial delay on Application start, but ContentProviders may start before it, and require special attention).
